clang/runtime/dpct-rt/include/dpct/dpl_extras/dpcpp_extensions.h

```diff
@@ -441,22 +441,28 @@ class radix_sort {
       unsigned_keys[i] = detail::traits<T>::twiddle_in(unsigned_keys[i]);
     }
 
-    while (true) {
+    for (int i = begin_bit; i < end_bit; i += RADIX_BITS) {
       int pass_bits = sycl::min(RADIX_BITS, end_bit - begin_bit);
 
       int ranks[VALUES_PER_THREAD];
       detail::radix_rank<RADIX_BITS, DESCENDING>(_local_memory)
-          .template rank_keys(item, unsigned_keys, ranks, begin_bit, pass_bits);
-      begin_bit += RADIX_BITS;
+          .template rank_keys(item, unsigned_keys, ranks, i, pass_bits);
 
       item.barrier(sycl::access::fence_space::local_space);
 
+      if (is_striped && i >= end_bit) {
```

Comment:
With the way the for loop is written, this condition will never be met. When `i` is incremented to be `>= end_bit` in the loop expression, then the condition will not be met and we will terminate the loop.

What you want to do here is modify this check to catch the last iteration.

Reply:
I believe the fix here does not correctly resolve the issue and instead adds an extra iteration. For example, if RADIX_BITS is 4 and you are are sorting a 32-bit type, then you will have 8 iterations in your loop. The recent changes to the for loop will falsely cause 9 iterations with `for (int i = begin_bit; i <= end_bit; i += RADIX_BITS)`.

I would recommend you revert the changes to the for loop and follow what @danhoeflinger suggested for the last iteration.